### PR TITLE
tikv: send heartbeat to idle store to detect liveness

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

we need a way to detect inactive store's liveness, and clean up them from region cache quickly if it has die

### What is changed and how it works?

send a heartbeat require to idle store, remove connection and invalid  region cache when heartbeat fail, TiClient will add heartbeat request and strip heartbeat response just like a interceptor pattern so upper logic(region-cache/cop/2pc..) never see heartbeat response.

### Check List 

Tests 

 - Unit test
 - Integration test(WIP)

Code changes

- send heartbeat if connection idle
- make `die` when heartbeat fail
- use hook to notify region-cache clean up die node 

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch(need more test)

Release note

 - send heartbeat to idle store to detect liveness

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/14145)

### CODE REVIEW

File: store/tikv/client_batch.go [hunk 1]
```diff
@@ -225,18 +234,22 @@ func (c *batchCommandsClient) send(request *tikvpb.BatchCommandsRequest, entries
...
 			zap.Error(err),
 		)
-		c.failPendingRequests(err)
-		return
+		return c.failPendingRequests(err)
+	}
+	if len(entries) == 1 {
+		failpoint.InjectContext(entries[0].ctx, "failBeforeSend", func() {
+			failpoint.Return(c.failPendingRequests(errors.New("test err")))
+		})
```
Comment (lonng) [inline store/tikv/client_batch.go]: Better to
```
		failpoint.InjectContext(entries[0].ctx, "failBeforeSend", func() bool {
			return c.failPendingRequests(errors.New("test err"))
		})
```

File: store/tikv/client_batch.go [hunk 2]
```diff
@@ -248,16 +261,25 @@ func (c *batchCommandsClient) recv() (*tikvpb.BatchCommandsResponse, error) {
...
+		heartbeatFail bool
+		i             int
+	)
 	c.batched.Range(func(key, value interface{}) bool {
 		id, _ := key.(uint64)
 		entry, _ := value.(*batchCommandsEntry)
 		entry.err = err
 		c.batched.Delete(id)
 		close(entry.res)
+		if i == 0 && entry.heartbeat {
```
Comment (tiancaiamao) [inline store/tikv/client_batch.go]: Will the `heartBeatEntry` always be the first element?
I see there is a `fetchMorePendingRequests` function.

**Comment by tiancaiamao** [comment]
LGTM

File: store/tikv/client_batch.go [hunk 3]
```diff
@@ -36,39 +36,39 @@ import (
 )
 
 type batchConn struct {
-	// An atomic flag indicates whether the batch is idle or not.
-	// 0 for busy, others for idle.
-	idle uint32
+	// An atomic flag indicates whether the batch is die or not.
+	// 0 for busy, others for die.
```
Comment (coocood) [inline store/tikv/client_batch.go]: Only two states die and busy?

File: store/tikv/client.go [hunk 4]
```diff
@@ -212,10 +214,19 @@ type rpcClient struct {
...
 	// Implement background cleanup.
-	isClosed bool
+	isClosed    bool
+	dieListener func(addr []string)
+}
+
+var _ ListenableRPCClient = &rpcClient{}
+
+// ListenableRPCClient present a RPCClient can be listen.
+type ListenableRPCClient interface {
```
Comment (coocood) [inline store/tikv/client.go]: I think we don't need to add this abstraction, we can check if it is `*rpcClient` directly.

File: store/tikv/client.go [hunk 5]
```diff
@@ -212,10 +214,11 @@ type rpcClient struct {
 	conns    map[string]*connArray
 	security config.Security
 
-	idleNotify uint32
-	// Periodically check whether there is any connection that is idle and then close and remove these idle connections.
+	dieNotify uint32
+	// Periodically check whether there is any connection that is die and then close and remove these connections.
```
Comment (siddontang) [inline store/tikv/client.go]: seem we should use "was died" instead of "is die"

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**